campaignion_tracking/js/campaignion_tracking.js

```diff
@@ -19,28 +19,58 @@ Drupal.behaviors.campaignion_tracking.attach = function(context, settings) {
       'data': data,
       'context': trackingContext
     }
-    window.campaignion_tracking.tracker.publish('donation', msg)
+    window.campaignion_tracking.tracker.publish(type, msg)
   }
 
-  /**
-   * Donation related events.
-   *
-   * The events are only fired if a correct campaignion_tracking context is
-   * available. They are only fired once for the Drupal context to prevent the
-   * same messages being dispatch multiple times (behaviours can be called
-   * multiple times).
-   *
-   * Events:
-   * - 'setDonationProduct'
-   * - 'checkoutBegin'
-   * - 'checkoutEnd'
-   */
   if (settings['campaignion_tracking'] && settings['campaignion_tracking']['context']) {
     settings['campaignion_tracking']['sent'] = settings['campaignion_tracking']['sent'] || [];
 
     var donation = settings.campaignion_tracking.context['donation'] || {};
     var webform = settings.campaignion_tracking.context['webform'] || {};
 
+    /**
+     * Webform related events.
+     *
+     * The events are only fired if a correct campaignion_tracking context is
+     * available. They are only fired once for the Drupal context to prevent the
+     * same messages being dispatch multiple times (behaviours can be called
+     * multiple times).
+     *
+     * Events:
+     * - 'draftBegin'
+     * - 'draftContinue'
+     */
+
+    // Fire `draftBegin` after completing the first step of multi-step forms.
+    if (!settings['campaignion_tracking']['sent'].includes('draftBegin')) {
+      if (webform['last_completed_step'] === 1) {
+        settings['campaignion_tracking']['sent'].push('draftBegin');
```

Comment:
Using Drupal.settings to store state is very weird it’s supposed to be only used for reading settings passed by the backend. Could `this` be used instead?

Reply:
No, the data needs to be persistent between behavior calls. An alternative might be to use the session storage (this could also help with comparing products as discussed in a previous comment), but that might not be so simple in a Drupal behavior?

Reply:
I think sessionStorage actually works quite well!

Reply:
`Drupal.behaviors.campaignion_tracking` has the same life time as `Drupal.settings`, so it would survive across AJAX calls but not across page loads.

Would the sessionStore need deletion when the donation is completed?

Reply:
Hm, removing the storage data would be a good idea, otherwise the events are only triggered on the first form of a session. But when should we do that, how can the behavior know the form is complete or a different form than before? :thinking:

Reply:
Shouldn’t the tracking work without AJAX enabled on forms as well? (Up until now this would have just triggered events too often.)

Reply:
What if someone does a one-time donation and a regular one with the same form? But yes, I’m not even sure what the behaviour should be.

Reply:
It would sent a "remove from cart" event for the one-time donation, a "add to cart" event for the regular one and another "purchase" event, but not another "begin checkout" or "end checkout".